hokusai/services/command_runner.py

```diff
@@ -1,71 +1,177 @@
+  def _image_name(self, tag_or_digest):
+    ''' generate docker image name '''
+    separator = '@' if ':' in tag_or_digest else ':'
+    image_name = f'{self.ecr.project_repo}{separator}{tag_or_digest}'
+    return image_name
 
-    container = {
-      "args": cmd.split(' '),
-      "name": name,
-      "image": image_name,
-      "imagePullPolicy": "Always",
-      'envFrom': [{'configMapRef': {'name': "%s-environment" % config.project_name}}]
-    }
+  def _set_env(self, container_spec, env):
+    ''' set env field of given container spec '''
+    spec = copy.deepcopy(container_spec)
+    spec['env'] = []
+    for var in env:
+      validate_key_value(var)
+      split = var.split('=', 1)
+      spec['env'].append(
+        {'name': split[0], 'value': split[1]}
+      )
+    return spec
 
-    run_tty = tty if tty is not None else config.run_tty
-    if run_tty:
-      container.update({
-        "stdin": True,
-        "stdinOnce": True,
-        "tty": True
-      })
+  def _set_envfrom(self, container_spec):
+    ''' set envFrom field of given container spec '''
+    spec = copy.deepcopy(container_spec)
+    spec.update(
+      {
+        'envFrom': [
+          {
+            'configMapRef': {
+              'name': f'{config.project_name}-environment'
+            }
+          },
+          {
+            'secretRef': {
+              'name': f'{config.project_name}',
+              'optional': True
+            }
+          }
+        ]
+      }
+    )
+    return spec
 
-    if env:
-      container['env'] = []
-      for s in env:
-        if '=' not in s:
-          raise HokusaiError("Error: environment variables must be of the form 'KEY=VALUE'")
-        split = s.split('=', 1)
-        container['env'].append({'name': split[0], 'value': split[1]})
-
-    spec = { "containers": [container] }
-    constraints = constraint or config.run_constraints
-    if constraints:
+  def _set_constraint(self, containers_spec, constraint):
+    ''' set nodeSelector field of given containers spec '''
+    spec = copy.deepcopy(containers_spec)
+    constraint = constraint or config.run_constraints
+    if constraint:
       spec['nodeSelector'] = {}
-      for label in constraints:
-        if '=' not in label:
-          raise HokusaiError("Error: Node selectors must of the form 'key=value'")
+      for label in constraint:
+        validate_key_value(label)
         split = label.split('=', 1)
         spec['nodeSelector'][split[0]] = split[1]
+    return spec
+
+  def _overrides_container(self, cmd, env, tag_or_digest):
+    ''' generate overrides['spec']['containers'][0] spec '''
+    spec = {
+      'args': cmd.split(' '),
+      'name': self.container_name,
+      'image': self._image_name(tag_or_digest),
+      'imagePullPolicy': 'Always',
+    }
+    spec = self._set_env(spec, env)
+    spec = self._set_envfrom(spec)
+    return spec
+
+  def _overrrides_containers(self, cmd, env, tag_or_digest):
```

Comment:
There's a typo here (extra r in overrrides) but I also don't think you need this function.

Reply:
good catch on the typo. the tests didn't catch this because they have the typo too!

Reply:
removed function